types/q/index.d.ts

```diff
@@ -145,7 +145,7 @@ declare namespace Q {
 		/**
 		 * A sugar method, equivalent to promise.then(function () { throw reason; }).
 		 */
-		thenReject(reason?: any): Promise<T>;
+		thenReject<U>(reason?: any): Promise<U>;
```

Comment:
Based on this comment:
> Looks good to me. This could potentially break some existing code so I would have liked to seen a solution that accommodated either but as I don't use this method and that it would be trivial to fix existing code. I'm OK with it.

You should write `<U = T>` instead of just `<U>` as the type parameter here, to avoid breaking existing consumers. `q` already declares as depending on TS 2.3+, so generic defaults should be available.

Reply:
I've tried adding a couple of new tests to try out what would happen for code that doesn't provide a type argument:

```
function thenRejectSameType(arg: any): Q.Promise<number> {
	if (!arg) {
		return returnsNumPromise('')
			.thenReject(new Error('failed'));
	}
	return Q.resolve(2);
}

// Test thenReject with a different type to the type of the preceding promise
function thenRejectOtherType(arg: any): Q.Promise<string> {
	if (!arg) {
		return returnsNumPromise('')
			.thenReject(new Error('failed'));
	}
	return Q.resolve('');
}

function thenRejectSpecificType(arg: any): Q.Promise<string> {
	if (!arg) {
		return returnsNumPromise('')
			.thenReject<string>(new Error('failed'));
	}
	return Q.resolve('');
}
```

With the definition I've already pushed, TypeScript 2.4 and up are happy. TS 2.3 reports `Type 'Promise<{}>' is not assignable to type 'Promise<number>'` for both `thenRejectSameType` and `thenRejectOtherType`. I think this is due to 'Return types as inference targets', added in 2.4.

Adding a default `U=T` to `thenReject` fixes the problem for TS 2.3 in `thenRejectSameType`, but it doesn't work in `thenRejectOtherType`, with the type argument not specified. 2.4+ is still happy.

This means that TS 2.3 users would have to specify the type argument if they need a different type, but existing code (which requires the same type) would work fine. TS 2.4+ users can omit the type argument.

I'm not quite sure what to do with the `thenRejectOtherType` test, omitting the type argument, since it fails for 2.3. `dtslint` doesn't allow us to flag that the error is expected for a particular version. I could just comment out or delete that test, but I would like to indicate to users that they can omit the type argument for TS 2.4+.

Reply:
> I'm not quite sure what to do with the thenRejectOtherType test, omitting the type argument, since it fails for 2.3. dtslint doesn't allow us to flag that the error is expected for a particular version. I could just comment out or delete that test, but I would like to indicate to users that they can omit the type argument for TS 2.4+.

In the end, I think you should comment out the test and leave a comment next to it to the effect of what you've said here. (And maybe a note to uncomment it if the version is ever bumped above 2.3)